File: strings/python/longest_substring_without_repeat.py
from typing import Optional


def longest_substring(s: Optional[str]) -> Optional[str]:
    """https://leetcode.com/problems/longest-substring-without-repeating-characters/"""
    if not s:
        return None
    visited = {}
    start = finish = max_length = tmp_start = 0
    for i in range(len(s)):
        if s[i] in visited and start <= visited[s[i]]:
            tmp_start = visited[s[i]] + 1
        else:
            tmp_length = i - tmp_start + 1
            if max_length < tmp_length:
                finish = i
                start = tmp_start
                max_length = tmp_length
        visited[s[i]] = i
    return s[start: finish + 1]  # len(s[start: finish + 1])


print(longest_substring('abcabcbb'))  # abc
print(longest_substring('bbbbb'))  # b
print(longest_substring('pwwkew'))  # wke

File: strings/go/longest_substring_without_repeat.go
package main

import (
	"fmt"
	"strings"
)

func longestSubstringWithoutRepeat(s string) (res string) {
	// https://leetcode.com/problems/longest-substring-without-repeating-characters/
	if s == "" {
		return res
	}
	sSlice := strings.Split(s, "")
	start, finish, maxLength, tmpStart := 0, 0, 0, 0
	visited := make(map[string]int)
	for i := range s {
		if val, ok := visited[sSlice[i]]; ok && start <= val {
			tmpStart = visited[sSlice[i]] + 1
		} else {
			tmpLength := i - tmpStart + 1
			if maxLength < tmpLength {
				finish = i
				start = tmpStart
				maxLength = tmpLength
			}
		}
		visited[sSlice[i]] = i
	}
	return s[start : finish+1]
}

func main() {
	fmt.Println(longestSubstringWithoutRepeat("abcabcbb")) // abc
	fmt.Println(longestSubstringWithoutRepeat("bbbbb"))    // b
	fmt.Println(longestSubstringWithoutRepeat("pwwkew"))   // "wke"
}
